### PR TITLE
zero out DiceData after reading it and keep state

### PR DESCRIPTION
**Pull request by jul-sh**
addresses https://github.com/project-oak/oak/pull/4377#discussion_r1361345702

I'm unsure if this is the right way.

Pro: I think zeroing out the private key is a good idea, and I think it should be done automatically once its been read.

Con: The stateful behavior of this descriptor is different from all other ones. They're stateless and always read from 0

### CODE REVIEW

**Review by conradgrobler — APPROVED** [review]
Thanks. Could you also please update the sealing key fd logic to do the same in a follow-up PR?

File: oak_restricted_kernel/src/syscall/dice_data.rs [hunk 1]
```diff
@@ -22,13 +22,24 @@ use oak_restricted_kernel_interface::{Errno, DICE_DATA_FD};
...
 }
 
 impl FileDescriptor for DiceDataDescriptor {
     fn read(&mut self, buf: &mut [u8]) -> Result<isize, oak_restricted_kernel_interface::Errno> {
-        let data_as_bytes = <DiceData as zerocopy::AsBytes>::as_bytes(&self.data);
-        let length = min(data_as_bytes.len(), buf.len());
-        buf.copy_from_slice(&data_as_bytes[..length]);
+        let data_as_bytes = <DiceData as zerocopy::AsBytes>::as_bytes_mut(&mut self.data);
+        let length = min(data_as_bytes.len() - self.index, buf.len());
+        let slice_to_read = &mut data_as_bytes[self.index..length];
```
Comment (conradgrobler) [inline oak_restricted_kernel/src/syscall/dice_data.rs:38]: Looking at it again, I don't think this works. I think it should be `[self.index..(self.index+length)]`, or the previous line could be changed to `let end = min(data_as_bytes.lent(), self.index + buf.len));`, and then this just becomes `[self.index..end]`

**Comment by conradgrobler** [comment]
> Thanks. Could you also please update the sealing key fd logic to do the same in a follow-up PR?

Actually, ignore the request. I think the ability to read a sealing key multiple times is fine, and perhaps needed.

Also, thinking about this a bit more: since the application is the last DICE layer zeroing out the keys are probably not an absolute requirement. If it is in an earlier layer you have to be sure that the key is destroyed before you launch the next layer, but in this case there are no later layer.

Regarding the pros and cons: I don't think we should force an application to know the size of the total struct, but might still be fine for an initial version. If we don't zero out the key we could zero the index when reaching the end so that an application can read the DICE data multiple times in multiple batches. Alternatively there is the option of adding yet another FD where the client can request the size of the DICE data, although that feels like the number of FDs will get out of hand.

So, up to you whether you want to back out of this change, submit as is (assuming my other comment about the length is addressed), or choose one of the other approaches.

**Comment by andrisaar** [comment]
+1 that another FD to read the length would feel really awkward and that's not how things are usually done.

I think maintaining some state (a "pointer" of how far we've read) is the simplest answer here; it'll mean that the FD behaves as you'd expect a FD on any other operating system to behave.

(Interesting question for the future: if we get to the point where we have multiple processes and threads in Restricted Kernel, how to handle concurrent reads?)

**Review by andrisaar — APPROVED** [review]

**Comment by jul-sh** [comment]
> I think maintaining some state (a "pointer" of how far we've read) is the simplest answer here; it'll mean that the FD behaves as you'd expect a FD on any other operating system to behave.`

that's a great point, I think it would be good to be consistent with how other FDs work. If we do with this way however, I think we should make the behavior consistent across all our FSs. At the moment some of them keep a pointer, others do not.

**Pull request merged**